src/Queue/BunnyProducer.php

```diff
@@ -0,0 +1,222 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Skrz\Bundle\BunnyBundle\Queue;
+
+use Bunny\Channel;
+use ReflectionClass;
+use Skrz\Bundle\BunnyBundle\Exception\BunnyException;
+use Skrz\Bundle\BunnyBundle\Service\BunnyManager;
+use Skrz\Bundle\BunnyBundle\Service\ContentTypes;
+use Symfony\Component\Serializer\SerializerInterface;
+
+/**
+ * @author Lukas Senfeld <skrz@senfeld.net>
+ * @method preProcessMessage(mixed $message, Channel $channel): void
+ */
+abstract class BunnyProducer implements BunnyProducerInterface
+{
+
+	private const NAME_SUFFIX_TO_REMOVE = "Producer";
+	private BunnyManager $manager;
+	private SerializerInterface $serializer;
+	private bool $autoCommit = false;
+	private string $exchange = "";
+	private bool $immediate = false;
+	private bool $mandatory = false;
+	private string $contentType = ContentTypes::APPLICATION_JSON;
+	private string $messageClassName;
+	private string $name;
+	private string $routingKey = "";
+	private bool $transactional = false;
+	private bool $hasBeforeMethod;
+
+	public function __construct(?string $name = null)
+	{
+		$this->name = $name ?? $this->generateName();
+		$this->hasBeforeMethod = method_exists($this, 'preProcessMessage');
+		$this->configure();
+	}
+
+	public function inject(BunnyManager $manager, SerializerInterface $serializer): void
+	{
+		$this->manager = $manager;
+		$this->serializer = $serializer;
+	}
+
+
+	/**
+	 * @param string|object $message
+	 * @param array<string,string> $headers
+	 */
+	public function publish($message, ?string $routingKey = null, array $headers = []): void
+	{
+		if (is_string($message)) {
+			$message = $this->serializer->deserialize($message, $this->messageClassName, 'json');
+		}
+
+		if ($this->hasBeforeMethod) {
```

Comment:
Napadá mě, že by se dal udělat (třeba) `BunnyPreProcessorInterface` s metodou `preProcessMessage`. Producer by ho pak implementoval a tady by se kontrolovalo `$this instanceof BunnyPreProcessorInterface`. Ale možná nevidím nějakou souvislost.